src/odemis/driver/piezomotor.py

```diff
@@ -0,0 +1,733 @@
+        logging.debug("Expecting a move of %g s, will wait up to %g s", dur, max_dur)
+
+        for ax, target in targets.items():
+            moving = True
+            end_time = time.time() + 5  # 5 s until timeout
+            while moving:
+                if time.time() > end_time:
+                    raise IOError("Timeout while waiting for end of motion on axis %s" % ax)
+                if self._closed_loop[ax]:
+                    if target is None:
+                        raise ValueError("No target provided for closed-loop move on axis %s." % ax)
+                    moving = self.isMovingClosedLoop(ax, target)
+                else:
+                    moving = self.isMovingOpenLoop(ax)
+                self._check_hw_error()
+                time.sleep(0.05)
+
+    def _check_hw_error(self):
+        for ax, axnum in self._axis_map.items():
+            status = self.getStatus(axnum)
+            if status[0] == 8:
+                raise PMDError(1, "Communication Error (wrong baudrate, data collision, or buffer overflow)")
+            elif status[0] == 4:
+                raise PMDError(2, "Encoder error(serial communication or reported error from serial encoder)")
+            elif status[0] == 2:
+                raise PMDError(3, "Supply voltage or motor fault was detected.")
+            elif status[1] == 1:
+                raise PMDError(4, "Command timeout occurred or a syntax error was detected when response was not allowed.")
+            elif status[1] == 8:
+                raise PMDError(5, "Power-on/reset has occurred.")
+
+    def _updatePosition(self):
+        """
+        update the position VA
+        """
+        pos = {}
+        for axname, axis in self._axis_map.items():
+            pos[axname] = self.getEncoderPosition(axis)
+        logging.debug("Reporting new position at %s", pos)
+        self.position._set_value(pos, force_write=True)
+
+    def stopAxis(self, axis):
+        self._sendCommand(b'X%dS' % self._axis_map[axis])
+
+    def getVersion(self, axis):
+        """
+        :returns (str): controller type and firmware version, e.g. 'PMD401 V13'
+        """
+        return self._sendCommand(b'X%d?' % axis)
+
+    def getSerialNumber(self, axis):
+        """
+        :returns (str): serial number
+        """
+        return self._sendCommand(b'X%dY42' % axis)
+
+    def getEncoderPosition(self, axis):
+        """
+        :returns (dict: str --> int): current position of axes as reported by encoders
+        """
+        return int(self._sendCommand(b'X%dE' % axis)) * ENCODER_STEPSIZE
+
+    def setWaveform(self, axis, wf):
+        """
+        :arg wf (waveform): waveform to set
+        """
+        if wf not in (WAVEFORM_DELTA, WAVEFORM_RHOMB, WAVEFORM_PARK):
+            raise ValueError("wf %s not a valid waveform" % wf)
+
+        self._sendCommand(b'X%dM%d' % (axis, wf))
+
+    def getStatus(self, axis):
+        """
+        :returns (str): 4-number status code
```

Comment:
Please, don't return a string, but 4 ints, which is easier to handle later (d1 to d4).